### PR TITLE
Ledger: Support OP_RETURN and multiple transaction outputs

### PR DESCRIPTION
**Pull request by EchterAgo**
The multiple transaction output code is taken from Electrum with a few minor modifications.

### CODE REVIEW

**Comment by cculianu** [comment]
@EchterAgo Thanks so much for this.  It's getting late here -- I will look over this tomorrow at the latest.

**Comment by EchterAgo** [comment]
That's fine. I'll likely be up before you tomorrow (timezone) , so I'll do some more testing and post the results. So far I've only tested on Windows with testnet and mainnet. Tomorrow I'll test on Linux.

File: plugins/hw_wallet/plugin.py [hunk 1]
```diff
@@ -51,3 +51,13 @@ def close_wallet(self, wallet):
         for keystore in wallet.get_keystores():
             if isinstance(keystore, self.keystore_class):
                 self.device_manager().unpair_xpub(keystore.xpub)
+
+def is_any_tx_output_on_change_branch(tx: Transaction):
+    if not tx.output_info:
```
Comment (cculianu) [inline plugins/hw_wallet/plugin.py]: Hmm.  It probably doesn't matter for the extant code you have here -- but to reduce the burden on future programmers, and not raise AttributeError here  -- would you mind doing a check for the attribute here, eg `if not getattr(tx, 'output_info', None):` instead?  

It seems `tx.output_info` isn't always defined for every Transaction instance (blame this codebase).  Just defensive programming...

Comment (cculianu) [reply plugins/hw_wallet/plugin.py]: (And yes, I realize in this callpath it will only ever be reached after `wallet.add_hw_info` runs.. :)

Comment (EchterAgo) [reply plugins/hw_wallet/plugin.py]: AFAICT it is defined for any TX for a hardware wallet in `add_hw_info`. I'll add the check though.

**Comment by EchterAgo** [comment]
Added the change to `is_any_tx_output_on_change_branch`

**Comment by EchterAgo** [comment]
Tested on Linux with Nano S / Nano as well.

**Comment by cculianu** [comment]
Thanks for testing it on Linux... Awesome!


About the other change:

Yeah it does indeed get defined in the `add_hw_info` so this call-path would mean it's always defined (but can be empty) -- your code was 100% correct.  Not a criticism of your code in the slightest!

My pet peeve with the code you did not write (that is, the wallet.py backend) makes code fragile to "sometimes" have attributes defined and have functions you call depend on them being defined but not have that "sometimes" bit be mentioned anywhere -- not in a comment -- not in anything.

You just have to "know". 

As I'm sure yo'd agree -- without your latest change: In the future some unsuspecting programmer might use the "is_any_tx_output_on_change" function thinking it always works as a generic solution and be surprised by its dependence on always being called after `add_hw_info`. 

Hopefully the very first line of the function will show a programmer in the future reading the code how the attribute it depends on to make that decision is not always defined... thus saving future programmers time when reading the code.

I know it's a nit -- thanks so much for updating the branch.

I could also update the Transaction class to at least always have this attribute as `{}`.. but -- as is the case with this codebase -- maybe even *that* breaks something else.. (I'd have to review that in order to make sure nothing else depends on it *not* being there to make some critical decision)! 

Thanks again.

I'll probably merge this in the next hour or two after breakfast. :)

It looks really good.  Good work man.

**Comment by EchterAgo** [comment]
Yea, I understand why it is better to do that way, I just copied and slightly adjusted this from Electrum. Sadly I'm not the best with dynamic programming languages, and especially because of things like this I personally hate them :)

**Comment by cculianu** [comment]
Ha ha -- I come from a C++ background myself. I can relate.  Dynamism has its plusses and minuses like with anything.  I find I have to do a lot of the work a compiler normally does for me when writing new code (which is kind of annoying) -- like making sure nothing has a typo.  My editor helps with that.

I spend a LOT of time reading over code I've written in Python to catch stuff  the compiler normally would have done for you in a more static language.

One benefit of that is you end up sometimes coming up with a better design as you re-read your code.

The drawback is.. it's a huge huge timesuck.

You did great work here and I appreciate your copying Electrum -- it's good when our two codebases dont' diverge much especially for auxilliary functionality such as this one.  When we can cleanly apply commits back and forth is ideal.  So I appreciate your sensibilities in that regard. 👍

**Comment by cculianu** [comment]
I think all Ledger users with sufficiently new devices will be super happy they can finally play with the big boys and sign OP_RETURN tx's.  This is not a small deal here.  THANKS!!

**Comment by cculianu** [comment]
Ok, looks good.  I just wanted to make sure I read over every line of code before merging is all.  Thanks again!!!

**Pull request merged**